sdk/python/lib/pulumi/resource.py

```diff
@@ -517,18 +520,22 @@ def _expand_providers(options: 'ResourceOptions'):
 
 def _collapse_providers(opts: 'ResourceOptions'):
     # If we have only 0-1 providers, then merge that back down to the .provider field.
-    providers: List['ProviderResource'] = opts.providers
+    providers: Union[Mapping[str, ProviderResource], List[ProviderResource], None] = opts.providers
     if providers is not None:
         provider_length = len(providers)
         if provider_length == 0:
             opts.providers = None
-        elif provider_length == 1:
+        elif isinstance(providers, list) and provider_length == 1:
             opts.provider = providers[0]
             opts.providers = None
         else:
             opts.providers = {}
-            for prov in providers:
-                opts.providers[prov.package] = prov
+            if isinstance(providers, list):
+                for prov in providers:
+                    opts.providers[prov.package] = prov
+            elif isinstance(providers, dict):
+                for key, prov in providers:
+                    opts.providers[key] = prov
```

Comment:
Aside - I have to imagine that this is fixes an actual bug in the implementation as well?

Reply:
Yes, it must. I imagine specifying provider maps is not a common use case though.

Reply:
Yes, it must. I imagine specifying provider maps is not a common use case though.